monai/apps/generation/maisi/networks/diffusion_model_unet_maisi.py

```diff
@@ -0,0 +1,415 @@
+            spatial_dims=spatial_dims,
+            in_channels=num_channels[-1],
+            temb_channels=new_time_embed_dim,
+            norm_num_groups=norm_num_groups,
+            norm_eps=norm_eps,
+            with_conditioning=with_conditioning,
+            num_head_channels=num_head_channels[-1],
+            transformer_num_layers=transformer_num_layers,
+            cross_attention_dim=cross_attention_dim,
+            upcast_attention=upcast_attention,
+            use_flash_attention=use_flash_attention,
+            dropout_cattn=dropout_cattn,
+        )
+
+        # up
+        self.up_blocks = nn.ModuleList([])
+        reversed_block_out_channels = list(reversed(num_channels))
+        reversed_num_res_blocks = list(reversed(num_res_blocks))
+        reversed_attention_levels = list(reversed(attention_levels))
+        reversed_num_head_channels = list(reversed(num_head_channels))
+        output_channel = reversed_block_out_channels[0]
+        for i in range(len(reversed_block_out_channels)):
+            prev_output_channel = output_channel
+            output_channel = reversed_block_out_channels[i]
+            input_channel = reversed_block_out_channels[min(i + 1, len(num_channels) - 1)]
+
+            is_final_block = i == len(num_channels) - 1
+
+            up_block = get_up_block(
+                spatial_dims=spatial_dims,
+                in_channels=input_channel,
+                prev_output_channel=prev_output_channel,
+                out_channels=output_channel,
+                temb_channels=new_time_embed_dim,
+                num_res_blocks=reversed_num_res_blocks[i] + 1,
+                norm_num_groups=norm_num_groups,
+                norm_eps=norm_eps,
+                add_upsample=not is_final_block,
+                resblock_updown=resblock_updown,
+                with_attn=(reversed_attention_levels[i] and not with_conditioning),
+                with_cross_attn=(reversed_attention_levels[i] and with_conditioning),
+                num_head_channels=reversed_num_head_channels[i],
+                transformer_num_layers=transformer_num_layers,
+                cross_attention_dim=cross_attention_dim,
+                upcast_attention=upcast_attention,
+                use_flash_attention=use_flash_attention,
+                dropout_cattn=dropout_cattn,
+            )
+
+            self.up_blocks.append(up_block)
+
+        # out
+        self.out = nn.Sequential(
+            nn.GroupNorm(num_groups=norm_num_groups, num_channels=num_channels[0], eps=norm_eps, affine=True),
+            nn.SiLU(),
+            zero_module(
+                Convolution(
+                    spatial_dims=spatial_dims,
+                    in_channels=num_channels[0],
+                    out_channels=out_channels,
+                    strides=1,
+                    kernel_size=3,
+                    padding=1,
+                    conv_only=True,
+                )
+            ),
+        )
+
+    def forward(
```

Comment:
I would suggest splitting the forward method into smaller functional chunks to make it easier to understand and maintain.
Such as:
```
class DiffusionModelUNetMaisi(nn.Module):
    def __init__(self, ...):
        super().__init__()
    def forward(self, x, timesteps, context=None, class_labels=None, top_region_index=None, bottom_region_index=None, spacing=None):
        emb = self._get_time_and_class_embedding(timesteps, class_labels)
        emb = self._get_input_embeddings(emb, top_region_index, bottom_region_index, spacing)
        h = self.conv_in(x)
        down_block_res_samples = [h]
        h = self._apply_down_blocks(h, emb, context, down_block_res_samples)
        h = self.middle_block(h, emb, context)
        h = self._apply_up_blocks(h, emb, context, down_block_res_samples)
        h = self.out(h)
        return h if isinstance(h, torch.Tensor) else torch.tensor(h)

    def _get_time_and_class_embedding(self, timesteps, class_labels):
        return emb

    def _get_input_embeddings(self, emb, top_index, bottom_index, spacing):
        return emb

    def _apply_down_blocks(self, h, emb, context, down_block_res_samples):
        return h

    def _apply_up_blocks(self, h, emb, context, down_block_res_samples):
        return h
```

Reply:
Thanks. I have updated this part. Please proceed.